runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java

```diff
@@ -51,7 +55,31 @@ public BoundedQueueExecutor(
             keepAliveTime,
             unit,
             new LinkedBlockingQueue<>(),
-            threadFactory);
+            threadFactory) {
+          @Override
+          protected void beforeExecute(Thread t, Runnable r) {
+            super.beforeExecute(t, r);
+            synchronized (this) {
+              if (activeCount.get() == maximumPoolSize - 1) {
```

Comment:
nit I'm always afraid of off-by-one errors, but I guess a `>=` would be preferred.